### PR TITLE
More API Doc Edits

### PR DESCRIPTION
**Pull request by jiajiechen**
* l2_normalization

* Activation

* BlockGrad (same as stop_gradient)

* MakeLoss

@nswamy @zackchase @Roshrini @mli @madjam @piiswrong

### CODE REVIEW

File: src/operator/l2_normalization-inl.h [hunk 1]
```diff
@@ -33,17 +33,18 @@ struct L2NormalizationParam : public dmlc::Parameter<L2NormalizationParam> {
...
-    "norm for each instance in the batch; this is the default mode. "
-    "If set to channel, this operator will compute a cross channel norm at "
-    "each position of each instance. If set to spatial, this operator will compute "
-    "a norm for each channel.");
+    .describe("Normalization Mode. By default it is set to ``instance``, and the "
+    "operator will compute an L2 norm for each instance in the multidimensional array. "
+    "If it is set to ``channel``, the operator will compute an L2 norm for each channel in "
+    "the multidimensional array. "
+    "If it is set to ``spatial``, the operator will compute a cross channel norm for each "
+    "position in the multidimensional array.");
```
Comment (mli) [inline src/operator/l2_normalization-inl.h]: suggest to move the long argument doc to the description of the function

File: src/operator/l2_normalization.cc [hunk 2]
```diff
@@ -19,8 +19,16 @@ Operator* L2NormalizationProp::CreateOperator(Context ctx) const {
...
 MXNET_REGISTER_OP_PROPERTY(L2Normalization, L2NormalizationProp)
-.describe("Set the l2 norm of each instance to a constant.")
-.add_argument("data", "NDArray-or-Symbol", "Input data to the L2NormalizationOp.")
+.describe(R"code(Normalize the input instance using a L2 norm.
+
+For 1-D NDArray, it computes::
+
+  out = data / sqrt(sum(data ** 2) + eps)
+
+The parameter ``mode`` can specify the dimension along which to compute L2 norm.
```
Comment (mli) [inline src/operator/l2_normalization.cc]: explain more about general n-d input

Comment (mli) [inline src/operator/l2_normalization.cc]: also add a particular example such as l2normalize([[1,2,3],[2,3,4]]) = [[...]]]

File: src/operator/make_loss.cc [hunk 3]
```diff
@@ -28,11 +28,29 @@ Operator *MakeLossProp::CreateOperatorEx(Context ctx, std::vector<TShape> *in_sh
...
+  \sum_i w_i * (y_i * \log \hat{y_i} + (1 - y_i) * \log(1 - \hat{y_i}))
+
+
+The following is a pseudocode snippet to create the customized loss::
+
+  y = Variable('y')
+  w = Variable('w')
+  out = Activation(data = data, act_type = 'sigmoid')
+  cross_entropy = y * log(out) + (1 - y) * log(1 - out)
+  loss = MakeLoss(w * cross_entropy)
```
Comment (mli) [inline src/operator/make_loss.cc]: change to something like:

assume `out` is the predicted output and `label` is the true label, then cross entropy can be defined by

  cross_entropy = label * log(out) + (1 - label) * log(1 - out)

now create the loss

   loss = MakeLoss(cross_entropy)

in additional, we can give a scale to loss by grad_scale=.1

also explain why Makeloss is needed, namely `stop_gradient` is called. also support `scale`

File: src/operator/tensor/elemwise_unary_op.cc [hunk 4]
```diff
@@ -29,7 +29,13 @@ NNVM_REGISTER_OP(_backward_copy)
 
 MXNET_OPERATOR_REGISTER_UNARY(BlockGrad)
 .add_alias("stop_gradient")
-.MXNET_DESCRIBE("Get output from a symbol and pass 0 gradient back")
+.MXNET_DESCRIBE(R"code(Stops gradient computation.
+
+Stops the accumulated gradient of the inputs from flowing through this operator
```
Comment (mli) [inline src/operator/tensor/elemwise_unary_op.cc]: explain more what it means, best by examples

File: src/operator/activation.cc [hunk 5]
```diff
@@ -66,15 +66,17 @@ DMLC_REGISTER_PARAMETER(ActivationParam);
 
 MXNET_REGISTER_OP_PROPERTY(Activation, ActivationProp)
 .describe(R"code(Elementwise activation function.
-The activation operations are applied elementwisely to each array elements. The
-following types are supported:
+The activation operations are applied elementwisely to each array elements.
```
Comment (zackchase) [inline src/operator/activation.cc]: Hi "elementwisely" occurred frequently throughout the old docs but it's not a word :). Let's say "element-wise" instead (the proper term and standard on numpy docs). Thanks!

File: src/operator/l2_normalization.cc [hunk 6]
```diff
@@ -19,8 +19,60 @@ Operator* L2NormalizationProp::CreateOperator(Context ctx) const {
...
+
+For N-D NDArray, if the input array has shape (N, N, ..., N),
+
+with ``mode`` = ``instance``, it normalizes each instance in the multidimensional
+array by its L2 norm.::
+
+  for i in 0...N
+    out[i,:,:,...,:] = data[i,:,:,...,:] / sqrt(sum(data[i,:,:,...,:] ** 2) + eps)
+
+with ``mode`` = ``channel``, it normalize each channel in the array by its L2 norm.::
```
Comment (zackchase) [inline src/operator/l2_normalization.cc]: Conjugation issue: should say "it normalizes" as above (not "it normalize")

File: src/operator/l2_normalization.cc [hunk 7]
```diff
@@ -19,8 +19,60 @@ Operator* L2NormalizationProp::CreateOperator(Context ctx) const {
...
+
+  for i in 0...N
+    out[:,i,:,...,:] = data[:,i,:,...,:] / sqrt(sum(data[:,i,:,...,:] ** 2) + eps)
+
+with ``mode`` = ``spatial``, it normalize the cross channel norm for each position
```
Comment (zackchase) [inline src/operator/l2_normalization.cc]: Conjugation issue: should say "it normalizes" as above (not "it normalize")

File: src/operator/make_loss-inl.h [hunk 8]
```diff
@@ -32,19 +32,19 @@ struct MakeLossParam : public dmlc::Parameter<MakeLossParam> {
   float valid_thresh;
   DMLC_DECLARE_PARAMETER(MakeLossParam) {
     DMLC_DECLARE_FIELD(grad_scale).set_default(1.0f)
-    .describe("gradient scale as a supplement to unary and binary operators");
+    .describe("Gradient scale as a supplement to unary and binary operators");
     DMLC_DECLARE_FIELD(valid_thresh).set_default(0.0f)
-    .describe("regard element valid when x > valid_thresh, this is "
-    "used only in valid normalization mode.");
+    .describe("Clip x to 0 when x <= valid_thresh, where x is the elements in the "
```
Comment (zackchase) [inline src/operator/make_loss-inl.h]: Might be more intuitive if you could just say "clip each element in the array to 0 when it is less than `valid_thresh`.

File: src/operator/tensor/elemwise_unary_op.cc (hunk 4 above)
Comment (jiajiechen) [inline src/operator/tensor/elemwise_unary_op.cc]: Edited. Preview as below.

screen shot 2017-04-20 at 1 32 19 pm

File: src/operator/make_loss.cc (hunk 3 above)
Comment (jiajiechen) [inline src/operator/make_loss.cc]: Edited, Preview as below

screen shot 2017-04-20 at 1 32 04 pm

File: src/operator/make_loss-inl.h (hunk 8 above)
Comment (jiajiechen) [inline src/operator/make_loss-inl.h]: Thanks! fixed!

File: src/operator/l2_normalization.cc (hunk 7 above)
Comment (jiajiechen) [inline src/operator/l2_normalization.cc]: Thanks! fixed!

File: src/operator/l2_normalization.cc (hunk 6 above)
Comment (jiajiechen) [inline src/operator/l2_normalization.cc]: Thanks! fixed!

File: src/operator/activation.cc (hunk 5 above)
Comment (jiajiechen) [inline src/operator/activation.cc]: Thanks! fixed!

File: src/operator/activation.cc [hunk 9]
```diff
@@ -65,16 +65,18 @@ Operator *ActivationProp::CreateOperatorEx(Context ctx, std::vector<TShape> *in_
 DMLC_REGISTER_PARAMETER(ActivationParam);
 
 MXNET_REGISTER_OP_PROPERTY(Activation, ActivationProp)
-.describe(R"code(Elementwise activation function.
-The activation operations are applied elementwisely to each array elements. The
-following types are supported:
+.describe(R"code(Element-wise activation function.
+The activation operations are applied element-wisely to each array elements.
```
Comment (madjam) [inline src/operator/activation.cc]: suggest changing "element-wisely to each array elements" to "element-wise to each array element"

**Comment by zackchase** [comment]
Madan is right. Please make this change!

File: src/operator/l2_normalization.cc [hunk 10]
```diff
@@ -19,8 +19,60 @@ Operator* L2NormalizationProp::CreateOperator(Context ctx) const {
 DMLC_REGISTER_PARAMETER(L2NormalizationParam);
 
 MXNET_REGISTER_OP_PROPERTY(L2Normalization, L2NormalizationProp)
-.describe("Set the l2 norm of each instance to a constant.")
-.add_argument("data", "NDArray-or-Symbol", "Input data to the L2NormalizationOp.")
+.describe(R"code(Normalize the input array using a L2 norm.
```
Comment (madjam) [inline src/operator/l2_normalization.cc]: suggest changing "a L2 norm" to "the L2 norm"

File: src/operator/activation.cc (hunk 9 above)
Comment (jiajiechen) [inline src/operator/activation.cc]: Thanks Madan! Fixed

File: src/operator/l2_normalization.cc (hunk 10 above)
Comment (jiajiechen) [inline src/operator/l2_normalization.cc]: Thanks! Fixed

**Comment by mli** [comment]
@yajiedesign can you have a look why win gpu is failed, it is caused by test_operator_gpu.test_convolution_dilated_impulse_response

**Comment by yajiedesign** [comment]
@mli some accidental errors
